contracts/peg/curve/CurveRouter.sol

```diff
@@ -0,0 +1,217 @@
+        );
+    }
+
+    // ---------- State-Changing API ----------
+
+    /// @notice Mint volt for stablecoins via curve
+    /// @param to, the address to mint Volt to
+    /// @param amountIn, the amount of stablecoin to deposit
+    /// @param amountVoltOut the amount of Volt we should get out, calculated externally from PSM and passed here
+    /// @param psm, the PSM the router should mint from
+    /// @param tokenA, the inital token that the user would like to swap
+    /// @param tokenB, the token the user would route through
+    /// @param noOfTokens, the number of tokens in the curve pool
+    /// @return amountOut the amount of Volt returned from the mint function
+    function mint(
+        address to,
+        uint256 amountIn,
+        uint256 amountVoltOut,
+        IPegStabilityModule psm,
+        address curvePool,
+        address tokenA,
+        address tokenB,
+        uint256 noOfTokens
+    ) external override returns (uint256 amountOut) {
+        (
+            uint256 amountTokenBReceived,
+            uint256 index_i,
+            uint256 index_j
+        ) = calculateSwap(amountIn, curvePool, tokenA, tokenB, noOfTokens);
+
+        IERC20(tokenA).transferFrom(msg.sender, address(this), amountIn);
+
+        ICurvePool(curvePool).exchange(
+            index_i.toInt256().toInt128(),
+            index_j.toInt256().toInt128(),
+            amountIn,
+            amountTokenBReceived
+        );
+
+        amountOut = psm.mint(to, amountTokenBReceived, amountVoltOut);
+    }
+
+    /// @notice Redeems volt for stablecoin via curve
+    /// @param to, the address to send redeemed stablecoin to
+    /// @param amountVoltIn, the amount of VOLT to deposit
+    /// @param minAmountOut, the minimum amount of stablecoin expected to be received
+    /// @param psm, the PSM the router should redeem from
+    /// @param tokenA, the token to route through on redemption
+    /// @param tokenB, the token the user would like to redeem
+    /// @return amountOut the amount of stablecoin returned from the mint function
+    function redeem(
+        address to,
+        uint256 amountVoltIn,
+        uint256 minAmountOut,
+        IPegStabilityModule psm,
+        address curvePool,
+        address tokenA,
+        address tokenB,
+        uint256 noOfTokens
+    ) external override returns (uint256) {
+        volt.transferFrom(msg.sender, address(this), amountVoltIn);
+
+        uint256 amountTokenAReceived = psm.redeem(
+            address(this),
+            amountVoltIn,
+            minAmountOut
+        );
+
+        (uint256 amountOut, uint256 index_i, uint256 index_j) = calculateSwap(
+            amountTokenAReceived,
+            curvePool,
+            tokenA,
+            tokenB,
+            noOfTokens
+        );
+
+        ICurvePool(curvePool).exchange(
+            index_i.toInt256().toInt128(),
+            index_j.toInt256().toInt128(),
+            amountTokenAReceived,
+            amountOut
```

Comment:
this should be calculated and passed from the UI to protect against front running